src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int length;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<T>(null, value, null);
+            tail = head;
+            tail.setPrev(head);
+            head.setNext(head);
+            length++;
+            return;
+        }
+        Node<T> newTail = new Node<T>(tail, value, null);
+        tail.setNext(newTail);
+        newTail.setPrev(tail);
+        tail = newTail;
+        length++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == length || (isEmpty() && index > -1)) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<T>(null, value, head);
+            head.setPrev(newNode);
+            head = newNode;
+        }
+        Node<T> after = getNode(index, true);
+        Node<T> before = after.getPrev();
+        Node<T> newNode = new Node<T>(before, value, after);
+        if (after != null) {
+            after.setPrev(newNode);
+        }
+        if (before != null) {
+            before.setNext(newNode);
+        }
```

Comment:
consider something like this:
```suggestion
        if (before == null) {
            head = newNode;
        }
        else{
            before.prev = newNode;
        }
```